include/revng/EarlyFunctionAnalysis/ControlFlowGraph.h

```diff
@@ -75,17 +75,22 @@ inline ParsedSuccessor parseSuccessor(const T &Edge,
 /// A function for converting EFA's internal CFG representation into a generic
 /// graph.
 ///
-/// \p BB An arbitrary container of basic blocks that are verified
-/// using the `IsBasicBlock` concept. These blocks are required to have start
-/// and end addresses as well as a list of their successors. It's expected for
-/// the graph it represents to be self contained, as in "no block can ever
-/// reference another block that is not listed in this container".
+/// \param BB An arbitrary container of basic blocks that are verified
+///        using the `IsBasicBlock` concept. These blocks are required to have
+///        a start and an end addresses as well as a list of their successors.
+///        It's expected for the graph it represents to be self contained, as
+///        in "no block can ever reference another block that is not listed in
+///        this container".
```

Comment:
Comment on the commit:

>  EFA: upgrade buildControlFlowGraph
>
> It now assumes that the nodes of the graph it constructs also require
> the function entry point on top of the basic block ID.

Why?
Later on we switch to `pipeline::Location` which makes this change obsolete. No fixups!
Squash?

Reply:
Ah, wait, no, I mixed pieces up. This change IS necessary exactly because we switch stuff relying on this to use locations later.

Before now we were using `BasicBlockID`s, so all we needed to build one was an ID. Now that the locations are built from the stuff this function returns, we also need to know the function entry point for each basic block we receive. Hence, this change.

Reply:
It could still be squashed, technically, since it's done for a similar purpose. BUT consider that this commit affects `efa` part of the codebase, while being the only commit to do so. As such, I'd rather keep it as self-contained as possible.

Reply:
1. Commit X: don't just use `BasicBlockID` but add function entry `MetaAddress`
2. Commit Y: drop `BasicBlockID` and `MetaAddress` in favor of `Location<rank::BasicBlock>` which embeds both

What prevents you from merging these two commits?

Reply:
It seems like you're misunderstanding the situation a bit.

The focus of the first commit is to make it so `buildControlFlowGraph` thing also throws a function address at the caller. (as of now it's a super-generic graph-builder that works with any `GenericGraph` you pass to it as long as its node is constructible from `{ BasicBlockID Block, MetaAddress Function }`)

And then the second commit does not really interact with it directly, it only reworks the internals of the graphs, but still keeps the constructor signature the same.

So, in essence, `X` is "patch **efa** to also give me function addresses" and `Y` is "update the **yield** graphs to store locations internally instead of whatever is used now".

Reply:
Gotcha, I'll check this again. I might have misunderstood.

Reply:
I was referring to the following change in the "EFA: upgrade buildControlFlowGraph" commit:

```patch
 struct Node {
   BasicBlockID Address;
   BasicBlockID NextAddress;

   explicit Node(BasicBlockID Address = BasicBlockID::invalid()) :
     Address(std::move(Address)) {}
+  explicit Node(BasicBlockID Address, const MetaAddress &) :
+    Address(std::move(Address)) {}
 };
```
Which is then basically completely overridden by a complete rewrite of the `struct Node` in "ControlFlow: switch to pipeline::Location".

Reply:
Ah, sorry for misunderstanding you then.

That's just an ad-hoc fix so that the code compiles after the change to the `efa`'s interface.

Reply:
I'm not sure I agree with "when changing an interface, all the users currently using it need to be update to use the new interface" qualifying as a trick.

The fact that we might want to rework some of these users later in the same PR doesn't feel like a reason enough to squash two logically-unrelated changes into a single commit.

Reply:
I guess that's a fair point.

I really didn't want to have to reorder these two commits (because then everything between them emits broken locations), but I guess it's still better than just squashing them.

Reply:
I appreciate your effort to make small commits and have a great history, but I'd like to get yourself tuned a lil0 bit more towards having larger commits.

Reply:
I'm doing it for a selfish reason: I often use `blame` view when looking at my old code to recover context. Having these things smaller means, well, reading less code to get the picture.